### PR TITLE
Adjust Histogram's bucket accounting to be iteratively

### PR DESCRIPTION
**Pull request by martijnvg**
Adjust Histogram's consumeBucketsAndMaybeBreak to be iteratively during reduce instead accounting all buckets at the end of the reduce.

In case of many non-empty buckets accounting the number of buckets at the end of the reduce may be too late. Elasticsearch may already have failed with an OOME. This change changes the accounting to happen iteratively during the reduce for non-empty bucket.

Note that for empty buckets accounting of the number of buckets already happens iteratively.

Similar to #101012

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalHistogram.java [hunk 1]
```diff
@@ -358,26 +368,14 @@ private double round(double key) {
...
-
     private void addEmptyBuckets(List<Bucket> list, AggregationReduceContext reduceContext) {
         /*
          * Make sure we have space for the empty buckets we're going to add by
          * counting all of the empties we plan to add and firing them into
          * consumeBucketsAndMaybeBreak.
          */
         class Counter implements DoubleConsumer {
-            private int size = list.size();
+            private int size = 0;
```
Comment (kkrik-es) [inline server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalHistogram.java:378]: Why is this changed?

**Review by kkrik-es — APPROVED** [review]

File: server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalHistogram.java (hunk 1 above)
Comment (martijnvg) [reply server/src/main/java/org/elasticsearch/search/aggregations/bucket/histogram/InternalHistogram.java:378]: This method only adds empty buckets. Before accounting of non empty buckets was either here or in the method calling this method of no empty buckets were requested. Now accounting of non empty buckets happens before this method is invoked. So `size` can be initialised with 0 instead of list.size() (this will contain non empty buckets if exists).

**Review by iverase — APPROVED** [review]
nice
